### ISSUE TITLE
About Token Merging Order

### ISSUE THREAD

**Issue by xiaokj37** [facebookresearch/ToMe#45]
Thank you very much for working open source.
Similar to issue#9, I would now like to be able to keep the original order as each block merges the token. 
However, as I look at your code implementation, I realize that you are splicing set A and B directly (sequentially，return torch.cat([unm, dst], dim=1) ), rather than alternating the choices. 
I would like to know, why is this? And how can I modify the implementation to keep the original order of the Token (the merged one can be put into set B).

**Comment by dbolya** [comment]
This is ablated in the paper--see Tab. 1e (Partitioning Style). Short answer is: ToMe uses bipartite matching. That means that not every token in each layer is compared to each other. By "mixing" the tokens as they are merged, we are guaranteeing that the next layer has the maximum number of new comparisons being made.

As for how to get the original order, supply `trace_source=True` to the patch function:
```
If you want to know the source of each token (e.g., for visualization), set trace_source = true.
The sources will be available at model._tome_info["source"] afterward.
```

**Comment by xiaokj37** [comment]
Thanks for your reply.
What I'm looking for is to keep the original Token order after each block, rather than getting it after the output of the model.
I tried the following modification, modifying the splice to an alternating merge.
`result_list = []
        unm_rows = unm.shape[1]
        dst_rows = dst.shape[1]

        max_rows = max(unm_rows, dst_rows)

        for i in range(max_rows):
            if i < unm_rows:
                result_list.append(unm[0, i, :])
            if i < dst_rows:
                result_list.append(dst[0, i, :])

        result = torch.stack(result_list).unsqueeze(0)`

**Comment by dbolya** [comment]
May I ask why you want this? It will destroy your accuracy.

FYI `model._tome_info["source"]` is populated every layer, so you could go down that route if you need intermediate features without destroying accuracy. If you really want to do this by tanking the algorithm's accuracy then the easiest way is probably just to not alternate at the start:

On line 51:
https://github.com/facebookresearch/ToMe/blob/main/tome/merge.py#L51

```py
a, b = metric[..., ::2, :], metric[..., 1::2, :]   ->   a, b = metric[..., :metric.shape[-2]//2, :], metric[..., metric.shape[-2]:, :]  
````

and on line 71:
https://github.com/facebookresearch/ToMe/blob/main/tome/merge.py#L71

```py
src, dst = x[..., ::2, :], x[..., 1::2, :]   ->   src, dst = x[..., :x.shape[-2]//2, :], x[..., x.shape[-2]//2:, :]
````

**Comment by xiaokj37** [comment]
I would like to be able to not merge the specified Token, similar to the class_token.
However, I found that the order changes after merging and I need to update the index of the specified Token.

**Comment by dbolya** [comment]
Is there just one of these tokens? If so, you can just make it the second token and set `model._tome_info["distill_token"] = True`. This will keep the second token always in the second spot and not merge it just like the class token.

Like here:
https://github.com/facebookresearch/ToMe/blob/af95e4b1befa172dadccd8c81e223b10090f9579/tome/patch/timm.py#L142-L143

If you need more than 2 tokens not merged, then I suggest setting their "metric"  to be -math.inf just like the others and keeping track of where they end up (or modify the algorithm to keep them in the first couple slots just like exists for cls and distill).

I heavily do not recommend keeping the tokens always in the same order, as like shown in the paper that will destroy accuracy.

**Comment by xiaokj37** [comment]
One thing I need to confirm is that the way you show in your paper, is it an “alternating” form, directly splicing the merged A and B?
<img width="344" alt="image" src="https://github.com/user-attachments/assets/d57b8fbc-4f03-4486-a998-b162d35f5275">
If so, I may have misunderstood what you meant in your paper.

**Comment by dbolya** [comment]
All forms directly concatenate A and B together. The difference is how A and B are constructed. 

![image](https://github.com/user-attachments/assets/f06905a6-e6a1-4727-819f-07f114da2caa)
Alternating does every other token into A and the rest into B.

Since the sets are constructed alternated but returned by concatenating, this shuffles the token order. If you constructed A as the first half of the tokens and B as the second half, then concatenating them at the end would not change their order (this is described as "sequential" in the paper).

**Comment by xiaokj37** [comment]
Thanks for your replay.
I'm now looking to track the original Token order by source, but I'm printing out the source matrix as following, can you help me explain this, thanks.
`tensor([[[0., 0., 1., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 1., 0., 0., 0., 0.],
         [0., 0., 0., 0., 1., 0., 0., 0., 0., 0., 0.],
         [1., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 1., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
         [0., 0., 0., 1., 0., 0., 0., 0., 1., 0., 0.],
         [0., 0., 0., 0., 0., 1., 0., 0., 0., 0., 1.],
         [0., 0., 0., 0., 0., 0., 0., 1., 0., 0., 0.],
         [0., 0., 0., 0., 0., 0., 0., 0., 0., 1., 0.]]])`

**Comment by xiaokj37** [comment]
Thanks to your patience, I have understood how to use model._tome_info[“source”].
Once again, thank you for your work.
